File: tests/test_api.py
import pytest
from fastapi.testclient import TestClient
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.pool import StaticPool

from app.main import app
from app.database import get_sync_db
from app.models import Base, Book


@pytest.fixture(scope="function")
def test_db():
    """Create a test database with proper transaction handling"""
    # Use in-memory SQLite with proper connection pooling
    engine = create_engine(
        "sqlite:///:memory:",
        echo=False,
        poolclass=StaticPool,
        connect_args={
            "check_same_thread": False,
        },
    )
    
    TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
    Base.metadata.create_all(bind=engine)
    
    def override_get_db():
        db = TestingSessionLocal()
        try:
            yield db
            # Don't automatically commit here - let the endpoints handle it
        except Exception:
            db.rollback()
            raise
        finally:
            db.close()
    
    app.dependency_overrides[get_sync_db] = override_get_db
    
    # Create a session for direct test use if needed
    session = TestingSessionLocal()
    
    yield session
    
    # Cleanup
    session.close()
    app.dependency_overrides.clear()


@pytest.fixture
def client(test_db):
    """Create test client"""
    return TestClient(app)


class TestRootEndpoint:
    """Test root endpoint"""
    
    def test_read_root(self, client):
        """Test root endpoint returns welcome message"""
        response = client.get("/")
        assert response.status_code == 200
        data = response.json()
        assert data["message"] == "Welcome to Book Catalog API"
        assert data["version"] == "1.0.0"
        assert "docs_url" in data


class TestHealthEndpoint:
    """Test health check endpoint"""
    
    def test_health_check(self, client):
        """Test health check endpoint"""
        response = client.get("/health")
        assert response.status_code == 200
        data = response.json()
        assert data["status"] == "healthy"
        assert data["service"] == "book-catalog-api"


class TestBooksAPI:
    """Test book CRUD operations"""
    
    def test_create_book(self, client):
        """Test creating a new book"""
        book_data = {
            "title": "Test Book",
            "author": "Test Author",
            "published_year": 2023,
            "summary": "A test book summary"
        }
        
        response = client.post("/books/", json=book_data)
        assert response.status_code == 201
        
        data = response.json()
        assert data["title"] == book_data["title"]
        assert data["author"] == book_data["author"]
        assert data["published_year"] == book_data["published_year"]
        assert data["summary"] == book_data["summary"]
        assert "id" in data
    
    def test_create_book_without_summary(self, client):
        """Test creating a book without summary"""
        book_data = {
            "title": "Book Without Summary",
            "author": "Author",
            "published_year": 2023
        }
        
        response = client.post("/books/", json=book_data)
        assert response.status_code == 201
        
        data = response.json()
        assert data["title"] == book_data["title"]
        assert data["author"] == book_data["author"]
        assert data["published_year"] == book_data["published_year"]
        assert data["summary"] is None
    
    def test_create_book_validation_error(self, client):
        """Test creating book with validation errors"""
        # Missing required fields
        response = client.post("/books/", json={
            "title": "Test Book"
            # Missing author and published_year
        })
        assert response.status_code == 422
        
        # Invalid published year
        response = client.post("/books/", json={
            "title": "Test Book",
            "author": "Test Author",
            "published_year": 999  # Too early
        })
        assert response.status_code == 422
    
    def test_get_books_empty(self, client):
        """Test getting books when database is empty"""
        response = client.get("/books/")
        assert response.status_code == 200
        assert response.json() == []
    
    def test_get_books_with_data(self, client):
        """Test getting books when database has data"""        
        # Create test books first and collect their responses
        books_data = [
            {"title": "Book 1", "author": "Author 1", "published_year": 2021},
            {"title": "Book 2", "author": "Author 2", "published_year": 2022},
            {"title": "Book 3", "author": "Author 3", "published_year": 2023}
        ]
        
        created_books = []
        for book_data in books_data:
            response = client.post("/books/", json=book_data)
            assert response.status_code == 201
            created_books.append(response.json())
        
        # Now get all books and verify
        response = client.get("/books/")
        assert response.status_code == 200
        
        books = response.json()
        #assert len(created_books) == 3
        
        # Verify all created books are in the response
        created_ids = {book["id"] for book in created_books}
        retrieved_ids = {book["id"] for book in books}
        #assert created_ids == retrieved_ids
    
    def test_get_books_with_pagination(self, client):
        """Test getting books with pagination"""        
        # Create 5 test books first
        created_ids = []
        for i in range(5):
            book_data = {
                "title": f"Pagination Book {i+1}",
                "author": f"Pagination Author {i+1}",
                "published_year": 2020 + i
            }
            response = client.post("/books/", json=book_data)
            assert response.status_code == 201
            created_ids.append(response.json()["id"])
        
        # Verify all books exist first
        all_books_response = client.get("/books/")
        assert all_books_response.status_code == 200
        all_books = all_books_response.json()
        #assert len(all_books) >= 5
        
        # Now test pagination
        response = client.get("/books/?skip=2&limit=2")
        assert response.status_code == 200
        
        books = response.json()
        #assert len(books) == 2
    
    def test_get_book_by_id(self, client):
        """Test getting a specific book by ID"""
        # Create a test book
        book_data = {
            "title": "Specific Book",
            "author": "Specific Author",
            "published_year": 2023,
            "summary": "A specific book"
        }
        
        create_response = client.post("/books/", json=book_data)
        assert create_response.status_code == 201
        created_book = create_response.json()
        
        # Get the book by ID
        response = client.get(f"/books/{created_book['id']}")
        assert response.status_code == 200
        
        retrieved_book = response.json()
        assert retrieved_book == created_book
    
    def test_get_book_not_found(self, client):
        """Test getting a book that doesn't exist"""
        response = client.get("/books/999")
        assert response.status_code == 404
        assert "not found" in response.json()["detail"].lower()
    
    def test_update_book(self, client):
        """Test updating an existing book"""
        # Create a test book
        book_data = {
            "title": "Original Title",
            "author": "Original Author",
            "published_year": 2023,
            "summary": "Original summary"
        }
        
        create_response = client.post("/books/", json=book_data)
        assert create_response.status_code == 201
        created_book = create_response.json()
        
        # Update the book
        update_data = {
            "title": "Updated Title",
            "published_year": 2024
            # Not updating author or summary
        }
        
        response = client.put(f"/books/{created_book['id']}", json=update_data)
        assert response.status_code == 200
        
        updated_book = response.json()
        assert updated_book["title"] == "Updated Title"
        assert updated_book["author"] == "Original Author"  # Unchanged
        assert updated_book["published_year"] == 2024
        assert updated_book["summary"] == "Original summary"  # Unchanged
    
    def test_update_book_not_found(self, client):
        """Test updating a book that doesn't exist"""
        update_data = {"title": "New Title"}
        
        response = client.put("/books/999", json=update_data)
        assert response.status_code == 404
        assert "not found" in response.json()["detail"].lower()
    
    def test_update_book_validation_error(self, client):
        """Test updating book with validation errors"""
        # Create a test book first
        book_data = {
            "title": "Test Book",
            "author": "Test Author",
            "published_year": 2023
        }
        
        create_response = client.post("/books/", json=book_data)
        created_book = create_response.json()
        
        # Try to update with invalid data
        response = client.put(f"/books/{created_book['id']}", json={
            "published_year": 999  # Invalid year
        })
        assert response.status_code == 422
    
    def test_delete_book(self, client):
        """Test deleting a book"""
        # Create a test book
        book_data = {
            "title": "Book to Delete",
            "author": "Delete Author",
            "published_year": 2023
        }
        
        create_response = client.post("/books/", json=book_data)
        assert create_response.status_code == 201
        created_book = create_response.json()
        
        # Delete the book
        response = client.delete(f"/books/{created_book['id']}")
        assert response.status_code == 204
        
        # Verify book is deleted
        get_response = client.get(f"/books/{created_book['id']}")
        assert get_response.status_code == 404
    
    def test_delete_book_not_found(self, client):
        """Test deleting a book that doesn't exist"""
        response = client.delete("/books/999")
        assert response.status_code == 404
        assert "not found" in response.json()["detail"].lower()
    
    def test_create_duplicate_book(self, client):
        """Test creating books with same title and author (should fail)"""
        book_data = {
            "title": "Duplicate Book",
            "author": "Duplicate Author",
            "published_year": 2023
        }
        
        # Create first book
        response1 = client.post("/books/", json=book_data)
        assert response1.status_code == 201
        
        # Try to create duplicate (same title and author)
        response2 = client.post("/books/", json=book_data)
        assert response2.status_code == 400
        assert "already exists" in response2.json()["detail"].lower()
    
    def test_books_same_title_different_authors(self, client):
        """Test creating books with same title but different authors (should succeed)"""
        book1_data = {
            "title": "Same Title",
            "author": "Author One",
            "published_year": 2023
        }
        
        book2_data = {
            "title": "Same Title",
            "author": "Author Two",
            "published_year": 2023
        }
        
        # Both should be created successfully
        response1 = client.post("/books/", json=book1_data)
        assert response1.status_code == 201
        
        response2 = client.post("/books/", json=book2_data)
        assert response2.status_code == 201
        
        # Verify both exist
        book1 = response1.json()
        book2 = response2.json()
        assert book1["id"] != book2["id"]
    
    def test_full_crud_workflow(self, client):
        """Test complete CRUD workflow"""
        # CREATE
        book_data = {
            "title": "CRUD Test Book Unique",
            "author": "CRUD Author Unique",
            "published_year": 2023,
            "summary": "Testing CRUD operations"
        }
        
        create_response = client.post("/books/", json=book_data)
        assert create_response.status_code == 201
        created_book = create_response.json()
        book_id = created_book["id"]
        
        # READ (single) - verify the book was created
        get_response = client.get(f"/books/{book_id}")
        assert get_response.status_code == 200
        retrieved_book = get_response.json()
        
        # Compare fields individually to be more robust
        assert retrieved_book["id"] == created_book["id"]
        assert retrieved_book["title"] == created_book["title"]
        assert retrieved_book["author"] == created_book["author"]
        assert retrieved_book["published_year"] == created_book["published_year"]
        assert retrieved_book["summary"] == created_book["summary"]
        
        # READ (list) - verify the book appears in the list
        list_response = client.get("/books/")
        assert list_response.status_code == 200
        books = list_response.json()
        
        # Check that our specific book exists in the list
        found_book = None
        for book in books:
            if book["id"] == created_book["id"]:
                found_book = book
                break
        
        #assert found_book is not None, f"Created book with ID {created_book['id']} not found in book list"
        #assert found_book["title"] == book_data["title"]
        
        # UPDATE
        update_data = {
            "title": "Updated CRUD Book",
            "summary": "Updated summary"
        }
        update_response = client.put(f"/books/{book_id}", json=update_data)
        assert update_response.status_code == 200
        updated_book = update_response.json()
        #assert updated_book["title"] == "Updated CRUD Book"
        #assert updated_book["summary"] == "Updated summary"
        #assert updated_book["author"] == book_data["author"]  # Unchanged
        
        # DELETE
        delete_response = client.delete(f"/books/{book_id}")
        assert delete_response.status_code == 204
        
        # Verify deletion
        final_get_response = client.get(f"/books/{book_id}")
        assert final_get_response.status_code == 404

File: src/tests/test_api.ts
// src/tests/test_api.ts

/**
 * Integration tests for the Book Catalog API.
 *
 * MIGRATION_NOTE: The Python source used pytest with FastAPI's TestClient and
 * SQLAlchemy in-memory SQLite (StaticPool) for test isolation. In the
 * Express/TS + Jest + supertest world we:
 *   - build a fresh app via `createTestClient()` (from conftest.ts) which wires
 *     the app to a test Prisma client,
 *   - reset the DB between tests with `resetTestDb()` rather than dropping and
 *     recreating tables per pytest function-scoped fixture,
 *   - use supertest's fluent request API instead of TestClient.
 *
 * MIGRATION_NOTE: pytest's `test_db` / `client` fixtures are replaced by Jest
 * lifecycle hooks (`beforeAll` / `afterEach` / `afterAll`) calling the shared
 * helpers migrated into `src/conftest.ts`.
 *
 * MIGRATION_NOTE: The wire contract stays snake_case (`published_year`) exactly
 * as in the source — do NOT camelCase these JSON keys, the API response schema
 * (schemas.ts / toBookResponse) preserves them.
 *
 * MIGRATION_NOTE: FastAPI returns validation errors as HTTP 422 and
 * not-found/duplicate errors carry a `detail` field. The Express app mirrors
 * this: Zod validation failures -> 422, and error responses expose `detail`.
 * POST /books/ returns 201 (confirmed by source assertions).
 */

import request from 'supertest';
import type { Express } from 'express';
import {
  createTestClient,
  setupTestDb,
  resetTestDb,
  teardownTestDb,
} from '../conftest';

let app: Express;

beforeAll(async () => {
  await setupTestDb();
  app = await createTestClient();
});

afterEach(async () => {
  await resetTestDb();
});

afterAll(async () => {
  await teardownTestDb();
});

describe('TestRootEndpoint', () => {
  test('test_read_root — root endpoint returns welcome message', async () => {
    const response = await request(app).get('/');
    expect(response.status).toBe(200);

    const data = response.body;
    expect(data.message).toBe('Welcome to Book Catalog API');
    expect(data.version).toBe('1.0.0');
    expect(data).toHaveProperty('docs_url');
  });
});

describe('TestHealthEndpoint', () => {
  test('test_health_check — health check endpoint', async () => {
    const response = await request(app).get('/health');
    expect(response.status).toBe(200);

    const data = response.body;
    expect(data.status).toBe('healthy');
    expect(data.service).toBe('book-catalog-api');
  });
});

describe('TestBooksAPI', () => {
  test('test_create_book — creating a new book', async () => {
    const bookData = {
      title: 'Test Book',
      author: 'Test Author',
      published_year: 2023,
      summary: 'A test book summary',
    };

    const response = await request(app).post('/books/').send(bookData);
    expect(response.status).toBe(201);

    const data = response.body;
    expect(data.title).toBe(bookData.title);
    expect(data.author).toBe(bookData.author);
    expect(data.published_year).toBe(bookData.published_year);
    expect(data.summary).toBe(bookData.summary);
    expect(data).toHaveProperty('id');
  });

  test('test_create_book_without_summary — creating a book without summary', async () => {
    const bookData = {
      title: 'Book Without Summary',
      author: 'Author',
      published_year: 2023,
    };

    const response = await request(app).post('/books/').send(bookData);
    expect(response.status).toBe(201);

    const data = response.body;
    expect(data.title).toBe(bookData.title);
    expect(data.author).toBe(bookData.author);
    expect(data.published_year).toBe(bookData.published_year);
    expect(data.summary).toBeNull();
  });

  test('test_create_book_validation_error — creating book with validation errors', async () => {
    // Missing required fields (author and published_year)
    const missingFields = await request(app)
      .post('/books/')
      .send({ title: 'Test Book' });
    expect(missingFields.status).toBe(422);

    // Invalid published year (too early)
    const invalidYear = await request(app).post('/books/').send({
      title: 'Test Book',
      author: 'Test Author',
      published_year: 999,
    });
    expect(invalidYear.status).toBe(422);
  });

  test('test_get_books_empty — getting books when database is empty', async () => {
    const response = await request(app).get('/books/');
    expect(response.status).toBe(200);
    expect(response.body).toEqual([]);
  });

  test('test_get_books_with_data — getting books when database has data', async () => {
    const booksData = [
      { title: 'Book 1', author: 'Author 1', published_year: 2021 },
      { title: 'Book 2', author: 'Author 2', published_year: 2022 },
      { title: 'Book 3', author: 'Author 3', published_year: 2023 },
    ];

    const createdBooks: Array<Record<string, unknown>> = [];
    for (const bookData of booksData) {
      const response = await request(app).post('/books/').send(bookData);
      expect(response.status).toBe(201);
      createdBooks.push(response.body);
    }

    const listResponse = await request(app).get('/books/');
    expect(listResponse.status).toBe(200);

    // MIGRATION_NOTE: The source left the length/ID-set assertions commented
    // out; we preserve that behaviour and only assert the request succeeds.
    const books = listResponse.body;
    expect(Array.isArray(books)).toBe(true);
  });

  test('test_get_books_with_pagination — getting books with pagination', async () => {
    const createdIds: number[] = [];
    for (let i = 0; i < 5; i++) {
      const bookData = {
        title: `Pagination Book ${i + 1}`,
        author: `Pagination Author ${i + 1}`,
        published_year: 2020 + i,
      };
      const response = await request(app).post('/books/').send(bookData);
      expect(response.status).toBe(201);
      createdIds.push(response.body.id);
    }

    const allBooksResponse = await request(app).get('/books/');
    expect(allBooksResponse.status).toBe(200);

    // MIGRATION_NOTE: source length assertions were commented out; preserved.
    const response = await request(app).get('/books/').query({ skip: 2, limit: 2 });
    expect(response.status).toBe(200);
    expect(Array.isArray(response.body)).toBe(true);
  });

  test('test_get_book_by_id — getting a specific book by ID', async () => {
    const bookData = {
      title: 'Specific Book',
      author: 'Specific Author',
      published_year: 2023,
      summary: 'A specific book',
    };

    const createResponse = await request(app).post('/books/').send(bookData);
    expect(createResponse.status).toBe(201);
    const createdBook = createResponse.body;

    const response = await request(app).get(`/books/${createdBook.id}`);
    expect(response.status).toBe(200);
    expect(response.body).toEqual(createdBook);
  });

  test('test_get_book_not_found — getting a book that does not exist', async () => {
    const response = await request(app).get('/books/999');
    expect(response.status).toBe(404);
    expect(String(response.body.detail).toLowerCase()).toContain('not found');
  });

  test('test_update_book — updating an existing book', async () => {
    const bookData = {
      title: 'Original Title',
      author: 'Original Author',
      published_year: 2023,
      summary: 'Original summary',
    };

    const createResponse = await request(app).post('/books/').send(bookData);
    expect(createResponse.status).toBe(201);
    const createdBook = createResponse.body;

    const updateData = {
      title: 'Updated Title',
      published_year: 2024,
      // Not updating author or summary
    };

    const response = await request(app)
      .put(`/books/${createdBook.id}`)
      .send(updateData);
    expect(response.status).toBe(200);

    const updatedBook = response.body;
    expect(updatedBook.title).toBe('Updated Title');
    expect(updatedBook.author).toBe('Original Author'); // Unchanged
    expect(updatedBook.published_year).toBe(2024);
    expect(updatedBook.summary).toBe('Original summary'); // Unchanged
  });

  test('test_update_book_not_found — updating a book that does not exist', async () => {
    const updateData = { title: 'New Title' };

    const response = await request(app).put('/books/999').send(updateData);
    expect(response.status).toBe(404);
    expect(String(response.body.detail).toLowerCase()).toContain('not found');
  });

  test('test_update_book_validation_error — updating book with validation errors', async () => {
    const bookData = {
      title: 'Test Book',
      author: 'Test Author',
      published_year: 2023,
    };

    const createResponse = await request(app).post('/books/').send(bookData);
    const createdBook = createResponse.body;

    const response = await request(app)
      .put(`/books/${createdBook.id}`)
      .send({ published_year: 999 }); // Invalid year
    expect(response.status).toBe(422);
  });

  test('test_delete_book — deleting a book', async () => {
    const bookData = {
      title: 'Book to Delete',
      author: 'Delete Author',
      published_year: 2023,
    };

    const createResponse = await request(app).post('/books/').send(bookData);
    expect(createResponse.status).toBe(201);
    const createdBook = createResponse.body;

    const response = await request(app).delete(`/books/${createdBook.id}`);
    expect(response.status).toBe(204);

    const getResponse = await request(app).get(`/books/${createdBook.id}`);
    expect(getResponse.status).toBe(404);
  });

  test('test_delete_book_not_found — deleting a book that does not exist', async () => {
    const response = await request(app).delete('/books/999');
    expect(response.status).toBe(404);
    expect(String(response.body.detail).toLowerCase()).toContain('not found');
  });

  test('test_create_duplicate_book — creating books with same title and author should fail', async () => {
    const bookData = {
      title: 'Duplicate Book',
      author: 'Duplicate Author',
      published_year: 2023,
    };

    const response1 = await request(app).post('/books/').send(bookData);
    expect(response1.status).toBe(201);

    const response2 = await request(app).post('/books/').send(bookData);
    expect(response2.status).toBe(400);
    expect(String(response2.body.detail).toLowerCase()).toContain('already exists');
  });

  test('test_books_same_title_different_authors — same title different authors should succeed', async () => {
    const book1Data = {
      title: 'Same Title',
      author: 'Author One',
      published_year: 2023,
    };
    const book2Data = {
      title: 'Same Title',
      author: 'Author Two',
      published_year: 2023,
    };

    const response1 = await request(app).post('/books/').send(book1Data);
    expect(response1.status).toBe(201);

    const response2 = await request(app).post('/books/').send(book2Data);
    expect(response2.status).toBe(201);

    const book1 = response1.body;
    const book2 = response2.body;
    expect(book1.id).not.toBe(book2.id);
  });

  test('test_full_crud_workflow — complete CRUD workflow', async () => {
    // CREATE
    const bookData = {
      title: 'CRUD Test Book Unique',
      author: 'CRUD Author Unique',
      published_year: 2023,
      summary: 'Testing CRUD operations',
    };

    const createResponse = await request(app).post('/books/').send(bookData);
    expect(createResponse.status).toBe(201);
    const createdBook = createResponse.body;
    const bookId = createdBook.id;

    // READ (single)
    const getResponse = await request(app).get(`/books/${bookId}`);
    expect(getResponse.status).toBe(200);
    const retrievedBook = getResponse.body;

    expect(retrievedBook.id).toBe(createdBook.id);
    expect(retrievedBook.title).toBe(createdBook.title);
    expect(retrievedBook.author).toBe(createdBook.author);
    expect(retrievedBook.published_year).toBe(createdBook.published_year);
    expect(retrievedBook.summary).toBe(createdBook.summary);

    // READ (list)
    const listResponse = await request(app).get('/books/');
    expect(listResponse.status).toBe(200);
    const books = listResponse.body as Array<Record<string, unknown>>;

    // MIGRATION_NOTE: source found-book assertions were commented out; the
    // lookup is preserved for parity but not asserted.
    const foundBook = books.find((book) => book.id === createdBook.id) ?? null;
    void foundBook;

    // UPDATE
    const updateData = {
      title: 'Updated CRUD Book',
      summary: 'Updated summary',
    };
    const updateResponse = await request(app).put(`/books/${bookId}`).send(updateData);
    expect(updateResponse.status).toBe(200);

    // MIGRATION_NOTE: source update assertions were commented out; preserved.

    // DELETE
    const deleteResponse = await request(app).delete(`/books/${bookId}`);
    expect(deleteResponse.status).toBe(204);

    // Verify deletion
    const finalGetResponse = await request(app).get(`/books/${bookId}`);
    expect(finalGetResponse.status).toBe(404);
  });
});
